### PR TITLE
ARROW-10037: [C++] Workaround to force find AWS SDK to look for shared libraries

### PR DESCRIPTION
**Pull request by kszucs**
Since the recent conda forge feedstock update find aws sdk fails to locate the proper aws sdk cmake files: https://github.com/apache/arrow/runs/1131824712#step:8:3700

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -2697,6 +2703,10 @@ if(ARROW_S3)
                             sts)
   endif()
 
+  if(DEFINED ENV{CONDA_PREFIX})
+    unset(BUILD_SHARED_LIBS)
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2707]: You're unsetting it even if it was already set.

Comment (kszucs) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2707]: I'm adding a flag for it then.

Comment (kszucs) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Updated.

**Review by xhochy — APPROVED** [review]
Lgtm, needs a format / autotune

**Comment by pitrou** [comment]
Uh... was Boost upgraded in the meantime? There are compile errors on AppVeyor:
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/35318362/job/lhejm2nbvp1rdtid

**Comment by kszucs** [comment]
Seems so, but that's a different issue.

**Comment by xhochy** [comment]
> Uh... was Boost upgraded in the meantime? There are compile errors on AppVeyor:
> https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/35318362/job/lhejm2nbvp1rdtid

As mentioned previously, the `boost-cpp[version='<1.70,>=1.68.0']` in AppVeyor is problematic. We build all Arrow dependencies currently with both 1.72 and 1.74, so both should work but not 1.68 or 1.69.

**Comment by pitrou** [comment]
Well, it would be nice if this could be fixed in this PR, otherwise we can't be sure that the fix works on Windows.

File: ci/conda_env_cpp.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@
 
 aws-sdk-cpp
 benchmark=1.4.1
-boost-cpp>=1.68.0
```
Comment (xhochy) [inline ci/conda_env_cpp.yml:20]: The lower-limit is no problem, I'm more confused why/where we have an upper limit. This will cause issues.

**Comment by pitrou** [comment]
```
CMake Error at C:/Miniconda37-x64/envs/arrow/Library/lib/cmake/Boost-1.74.0/BoostConfig.cmake:141 (find_package):
  Found package configuration file:
    C:/Miniconda37-x64/envs/arrow/Library/lib/cmake/boost_regex-1.74.0/boost_regex-config.cmake
  but it set boost_regex_FOUND to FALSE so package "boost_regex" is
  considered to be NOT FOUND.  Reason given by package:
  No suitable build variant has been found.
  The following variants have been tried and rejected:
  * boost_regex.lib (shared, Boost_USE_STATIC_LIBS=ON)
```

**Comment by pitrou** [comment]
Thank you @kszucs :-)

**Pull request closed without merging**